File: reportes/flotilla.py
"""
SIG-LOG — Sistema Integral de Gestión Logística
reportes/flotilla.py

INFORME DE FLOTILLA

Responde las cuatro preguntas del proyecto que se contestan mirando los
vehículos:

    ¿Qué vehículos generan mayores costos?
    ¿Qué vehículos consumen más combustible?
    ¿Qué vehículos tienen más entregas?
    ¿Qué vehículos presentan más retrasos?

Y una quinta que aparece al cruzarlas: qué unidades cuestan mucho para lo
poco que mueven. El costo total premia a las que trabajan menos, así que la
cifra que decide es el costo por entrega.

Cada unidad se identifica por su código y su modelo. El identificador
interno no aparece en ninguna parte del documento: a quien lee un informe
no le dice nada.
"""

from __future__ import annotations

import sys
from pathlib import Path
from typing import Any

RAIZ = Path(__file__).resolve().parents[1]
if str(RAIZ) not in sys.path:
    sys.path.insert(0, str(RAIZ))

from pymongo.database import Database

from backend.services import analitica, mantenimientos as servicio_mtto
from backend.utils.errores import ErrorSIGLOG
from config import settings
from reportes import comun

TITULO = "Informe de flotilla"
SUBTITULO = ("Desempeño de los vehículos: qué cuestan, qué consumen, cuánto "
             "trabajan y con qué puntualidad.")


def construir(bd: Database) -> bytes:
    datos = analitica.desempeno_vehiculos(bd, "costo", 100)
    totales = datos["totales"]
    periodo = datos["periodo"]
    umbral = datos["umbral_retraso_min"]
    flota = datos["vehiculos"]

    elementos: list = comun.portada(TITULO, SUBTITULO, periodo)

    # -------------------------------------------------------- el conjunto --
    elementos += comun.seccion(
        "1. La flotilla en conjunto",
        f"{datos['flotilla']} unidades en operación durante el periodo. "
        "Estas cifras son la referencia contra la que se juzga cada "
        "vehículo: una unidad cara solo lo es comparada con las demás.")

    elementos.append(comun.tabla(
        ["Costo de operación", "Combustible", "Distancia", "Entregas",
         "Rendimiento medio", "Fuera de operación"],
        [[comun.dinero(totales["costo_total"]),
          f"{comun.numero(totales['litros'], 0)} litros<br/>"
          f"<font size=6 color='#667589'>"
          f"{comun.dinero(totales['costo_combustible'])}</font>",
          f"{comun.numero(totales['km_recorridos'], 0)} km",
          f"{comun.entero(totales['entregas'])}<br/>"
          f"<font size=6 color='#667589'>en "
          f"{comun.entero(totales['viajes'])} viajes</font>",
          f"{comun.numero(totales['rendimiento_medio_km_l'], 2)} km/l",
          f"{comun.entero(totales['en_mantenimiento'])} unidades"]],
        alinear_derecha=(0, 1, 2, 3, 4, 5)))

    costo_medio_entrega = (totales["costo_total"] / totales["entregas"]
                           if totales["entregas"] else 0)
    elementos.append(comun.lectura(
        f"La operación de la flotilla costó "
        f"{comun.dinero(totales['costo_total'])} en el periodo, de los que "
        f"{comun.dinero(totales['costo_combustible'])} son combustible y "
        f"{comun.dinero(totales['costo_mantenimiento'])} mantenimiento. "
        f"Repartido entre las {comun.entero(totales['entregas'])} entregas "
        f"realizadas, sale a {comun.dinero(costo_medio_entrega, 2)} por "
        "entrega: esa es la cifra contra la que conviene comparar cada "
        "unidad."))

    # ------------------------------------------------------ las preguntas --
    elementos += _ranking(bd, "costo", "2. ¿Qué unidades cuestan más?",
                          "Combustible más mantenimiento acumulados en el "
                          "periodo. Una unidad cara no es necesariamente un "
                          "problema: puede ser la que más trabaja. Eso se "
                          "resuelve en el apartado 6.")

    elementos += _ranking(bd, "combustible",
                          "3. ¿Qué unidades consumen más combustible?",
                          "Litros cargados en el periodo. Consumir mucho es "
                          "esperable en una unidad que rueda mucho; el aviso "
                          "está en el rendimiento, no en el consumo.")

    elementos.append(comun.salto())

    elementos += _ranking(bd, "entregas", "4. ¿Qué unidades trabajan más?",
                          "Entregas realizadas. Junto al costo por entrega "
                          "es lo que permite saber si una unidad rinde lo "
                          "que cuesta.")

    elementos += _ranking(bd, "retraso", "5. ¿Qué unidades llegan tarde?",
                          "Retraso medio de sus entregas. El retraso rara "
                          "vez es culpa del vehículo: conviene cruzarlo con "
                          "las rutas que cubre antes de concluir nada.",
                          umbral=umbral)

    # ------------------------------------------------- costo frente a uso --
    elementos.append(comun.salto())
    elementos += _costo_frente_a_trabajo(flota, totales, datos["flotilla"])

    # ------------------------------------------------------- el detalle ---
    elementos.append(comun.salto())
    elementos += _detalle(flota, umbral)

    # ------------------------------------------------------ mantenimiento --
    elementos += _mantenimiento(bd)

    return comun.documento(TITULO, SUBTITULO, elementos)


# ==========================================================================
# BLOQUES
# ==========================================================================
_EJES = {
    "costo": ("costo_total", "Costo de operación (MXN)", "#d62728"),
    "combustible": ("litros", "Combustible consumido (litros)", "#ff7f0e"),
    "entregas": ("entregas", "Entregas realizadas", "#1f4e79"),
    "retraso": ("retraso_medio_min", "Retraso medio (minutos)", "#d62728"),
}


def _ranking(bd: Database, criterio: str, titulo: str, explicacion: str,
             umbral: int | None = None) -> list:
    datos = analitica.desempeno_vehiculos(bd, criterio, 8)
    filas = datos["vehiculos"]
    campo, eje, color = _EJES[criterio]

    def dibujar(ax):
        orden = list(reversed(filas))
        etiquetas = [v["codigo_vehiculo"] for v in orden]
        valores = [v[campo] or 0 for v in orden]
        colores = color
        if umbral is not None:
            colores = ["#d62728" if (v[campo] or 0) > umbral else "#1f4e79"
                       for v in orden]
        ax.barh(etiquetas, valores, color=colores, alpha=.9)
        if umbral is not None:
            ax.axvline(umbral, color="#8c9bb0", linestyle="--", linewidth=1.2)
            ax.text(umbral, -0.7, f"  umbral {umbral} min", fontsize=6,
                    color="#667589")
        ax.set_xlabel(eje)
        ax.set_ylabel("Vehículo")
        ax.set_title(titulo.split(". ", 1)[-1], fontsize=9, fontweight="bold")
        ax.grid(alpha=.25, axis="x")
        for i, valor in enumerate(valores):
            texto = (comun.dinero(valor) if criterio == "costo"
                     else comun.numero(valor, 0 if criterio != "retraso" else 1))
            ax.text(valor * 1.01, i, f" {texto}", va="center", fontsize=6)

    partes = comun.seccion(titulo, explicacion)
    partes.append(comun.grafica(dibujar, alto_cm=6.0))
    partes.append(comun.lectura(datos["lectura"]))
    return partes


def _costo_frente_a_trabajo(flota: list[dict[str, Any]],
                            totales: dict[str, Any], tamano: int) -> list:
    """
    El cruce que ninguna de las cuatro preguntas resuelve por separado.

    Una unidad cara que entrega mucho está bien; una unidad cara que entrega
    poco es donde hay dinero que recuperar.
    """
    costo_medio = totales["costo_medio_por_vehiculo"]
    entregas_media = totales["entregas"] / (tamano or 1)
    ineficientes = [v for v in flota
                    if v["costo_total"] > costo_medio
                    and v["entregas"] < entregas_media]

    def dibujar(ax):
        for grupo, color, etiqueta in (
            (ineficientes, "#d62728", "Cuesta más de lo que trabaja"),
            ([v for v in flota if v not in ineficientes], "#1f4e79",
             "Dentro de lo esperado"),
        ):
            ax.scatter([v["entregas"] for v in grupo],
                       [v["costo_total"] for v in grupo],
                       s=[30 + 170 * v["litros"] / (totales["litros"] or 1)
                          * tamano for v in grupo],
                       color=color, alpha=.6, label=etiqueta,
                       edgecolors="white", linewidth=.5)
        for v in ineficientes:
            ax.annotate(v["codigo_vehiculo"],
                        (v["entregas"], v["costo_total"]),
                        fontsize=6, xytext=(5, 4),
                        textcoords="offset points", color="#8a2020")
        ax.axhline(costo_medio, color="#8c9bb0", linestyle=":", linewidth=1)
        ax.axvline(entregas_media, color="#8c9bb0", linestyle=":", linewidth=1)
        ax.set_xlabel("Entregas realizadas")
        ax.set_ylabel("Costo de operación (MXN)")
        ax.set_title("Costo de operación frente a entregas realizadas",
                     fontsize=9, fontweight="bold")
        ax.legend(fontsize=7, loc="upper left")
        ax.grid(alpha=.2)

    partes = comun.seccion(
        "6. ¿Qué unidades cuestan más de lo que trabajan?",
        "Cada punto es un vehículo; el tamaño indica los litros consumidos. "
        "Las líneas punteadas marcan la media de la flotilla. Las unidades "
        "del cuadrante superior izquierdo —por encima del costo medio y por "
        "debajo de las entregas medias— son las primeras a revisar.")
    partes.append(comun.grafica(dibujar, alto_cm=8.0))

    if ineficientes:
        peor = max(ineficientes, key=lambda v: v["costo_por_entrega"] or 0)
        medio = (totales["costo_total"] / totales["entregas"]
                 if totales["entregas"] else 0)
        partes.append(comun.lectura(
            f"{len(ineficientes)} de las {tamano} unidades gastan por encima "
            f"de la media ({comun.dinero(costo_medio)}) y entregan por debajo "
            f"de ella ({comun.numero(entregas_media, 0)} entregas). La más "
            f"desproporcionada es {peor['codigo_vehiculo']} "
            f"({peor['descripcion']}), con "
            f"{comun.dinero(peor['costo_por_entrega'], 2)} por entrega frente "
            f"a los {comun.dinero(medio, 2)} del promedio de la flotilla."))
    else:
        partes.append(comun.lectura(
            "Ninguna unidad gasta por encima de la media entregando por "
            "debajo de ella: el costo de la flotilla acompaña al trabajo que "
            "hace cada vehículo."))
    return partes


def _detalle(flota: list[dict[str, Any]], umbral: int) -> list:
    partes = comun.seccion(
        "7. Detalle por unidad",
        "Todas las unidades, ordenadas por costo de operación. En rojo, las "
        f"que promedian más de {umbral} minutos de retraso o cuyo "
        "rendimiento real queda por debajo del de ficha.")

    filas = []
    for v in flota:
        filas.append([
            f"<b>{v['codigo_vehiculo']}</b><br/>"
            f"<font size=6 color='#667589'>{v['descripcion']}<br/>"
            f"{v['placa']}</font>",
            v["tipo_vehiculo"] or "—",
            comun.entero(v["viajes"]),
            comun.entero(v["entregas"]),
            comun.numero(v["km_recorridos"], 0),
            comun.dinero(v["costo_total"]),
            comun.dinero(v["costo_por_entrega"], 2),
            comun.numero(v["litros"], 0),
            f"{comun.numero(v['rendimiento_real_km_l'], 2)}<br/>"
            f"<font size=6 color='#667589'>ficha "
            f"{comun.numero(v['rendimiento_nominal_km_l'], 2)}</font>",
            (f"{comun.numero(v['retraso_medio_min'], 1)}<br/>"
             f"<font size=6 color='#667589'>"
             f"{comun.porcentaje(v['pct_retrasadas'])} tarde</font>"
             if v["retraso_medio_min"] is not None else "—"),
            comun.entero(v["mantenimientos"]),
        ])

    def problematica(i: int) -> bool:
        v = flota[i]
        tarde = (v["retraso_medio_min"] or 0) > umbral
        flojo = (v["desviacion_rendimiento_pct"] or 0) < 0
        return tarde or flojo

    partes.append(comun.tabla(
        ["Vehículo", "Tipo", "Viajes", "Entregas", "Km",
         "Costo total<br/>(MXN)", "Costo/entrega<br/>(MXN)",
         "Combustible<br/>(litros)", "Rendimiento<br/>(km/l)",
         "Retraso medio<br/>(minutos)", "Servicios"],
        filas,
        anchos=[comun.ANCHO_UTIL * p for p in
                (.15, .07, .06, .07, .08, .10, .09, .09, .10, .11, .08)],
        alinear_derecha=(2, 3, 4, 5, 6, 7, 8, 9, 10),
        destacar=problematica))
    return partes


def _mantenimiento(bd: Database) -> list:
    try:
        datos = servicio_mtto.pendientes(bd)
    except ErrorSIGLOG as error:
        return [comun.aviso(f"<b>Mantenimiento.</b> No disponible: "
                            f"{error.mensaje}")]

    partes = comun.seccion(
        "8. ¿Qué unidades requieren mantenimiento?",
        "Los servicios vencidos ya sacaron la unidad de operación y no "
        "puede programarse en una jornada. Los atrasados son los que hay "
        "que atender hoy. Los próximos aún se pueden planificar sin parar "
        "nada.")

    filas = []
    etiquetas = {"vencidos": "Vencido — unidad parada",
                 "atrasados": "Atrasado — atender hoy",
                 "proximos": "Próximo — planificable"}
    for grupo, etiqueta in etiquetas.items():
        for m in datos[grupo]:
            filas.append([
                f"<b>{m['codigo_vehiculo']}</b> "
                f"<font size=6 color='#667589'>{m.get('placa') or ''}</font>",
                etiqueta,
                m["tipo"] or "—",
                comun.fecha(m["fecha_programada"]),
                comun.entero(m["dias"]),
                (m["estado_operativo"] or "—").replace("_", " ").title(),
            ])

    if not filas:
        partes.append(comun.lectura(datos["alerta"]))
        return partes

    vencidos = datos["total_vencidos"]
    partes.append(comun.tabla(
        ["Vehículo", "Situación", "Tipo de servicio",
         "Fecha programada", "Días de diferencia", "Estado de la unidad"],
        filas,
        anchos=[comun.ANCHO_UTIL * p for p in
                (.16, .24, .14, .16, .15, .15)],
        alinear_derecha=(4,),
        destacar=lambda i: i < vencidos))
    partes.append(comun.lectura(datos["alerta"]))
    return partes

File: frontend/static/js/flotilla.js
/* =========================================================================
   SIG-LOG — desempeño de la flotilla

   Cada bloque de esta pantalla existe por una pregunta de negocio concreta.
   Ninguna cifra se calcula aquí: todo viene de /analitica/vehiculos, que a
   su vez lee `dim_vehiculo` y `hecho_entrega`.
   ========================================================================= */

(function () {
  "use strict";

  const UMBRAL = window.SIGLOG.umbral;
  const PERIODO = JSON.parse(
    document.getElementById("datos-periodo").textContent || "{}");

  // Cada criterio sabe cómo rotularse. Un eje sin unidad no se interpreta.
  const CRITERIOS = {
    costo: {
      campo: "costo_total", eje: "Costo de operación (MXN)",
      titulo: "Costo de operación por vehículo",
      formato: SL.dinero, color: SL.COLORES.alerta
    },
    combustible: {
      campo: "litros", eje: "Combustible consumido (litros)",
      titulo: "Combustible consumido por vehículo",
      formato: function (v) { return SL.numero(v, 0) + " l"; },
      color: SL.COLORES.aviso
    },
    entregas: {
      campo: "entregas", eje: "Entregas realizadas",
      titulo: "Entregas realizadas por vehículo",
      formato: SL.entero, color: SL.COLORES.principal
    },
    retraso: {
      campo: "retraso_medio_min", eje: "Retraso medio (minutos)",
      titulo: "Retraso medio por vehículo",
      formato: function (v) { return SL.numero(v, 1) + " min"; },
      color: SL.COLORES.alerta
    },
    rendimiento: {
      campo: "rendimiento_real_km_l", eje: "Rendimiento real (km/l)",
      titulo: "Rendimiento real por vehículo",
      formato: function (v) { return SL.numero(v, 2) + " km/l"; },
      color: SL.COLORES.bien
    },
    uso: {
      campo: "km_recorridos", eje: "Distancia recorrida (km)",
      titulo: "Kilometraje por vehículo",
      formato: function (v) { return SL.numero(v, 0) + " km"; },
      color: SL.COLORES.principal
    }
  };

  let grafica = null;
  let dispersion = null;
  let flotillaCompleta = [];

  /** Etiqueta legible de una unidad. El identificador interno no se enseña. */
  function etiqueta(v) {
    return v.codigo_vehiculo;
  }
  function etiquetaLarga(v) {
    return v.codigo_vehiculo + " · " + v.descripcion +
           (v.placa ? " (" + v.placa + ")" : "");
  }

  // ======================================================================
  // GRÁFICA PRINCIPAL
  // ======================================================================
  async function cargar() {
    const criterio = document.getElementById("criterio").value;
    const top = document.getElementById("cuantos").value;
    const cfg = CRITERIOS[criterio];

    try {
      const r = await SL.api("/analitica/vehiculos?orden=" + criterio +
                             "&top=" + top);
      const d = r.datos;
      const filas = d.vehiculos;
      document.getElementById("l-flotilla").textContent = d.lectura;

      if (grafica) grafica.destroy();
      grafica = new Chart(document.getElementById("g-flotilla"), {
        type: "bar",
        data: {
          labels: filas.map(etiqueta),
          datasets: [{
            label: cfg.eje,
            data: filas.map(function (v) { return v[cfg.campo]; }),
            backgroundColor: filas.map(function (v) {
              // El rojo señala lo que exige intervención, no el primer puesto
              if (criterio === "retraso") {
                return v.retraso_medio_min > UMBRAL
                  ? SL.COLORES.alerta : SL.COLORES.principal;
              }
              if (criterio === "rendimiento") {
                return (v.desviacion_rendimiento_pct || 0) < 0
                  ? SL.COLORES.alerta : SL.COLORES.bien;
              }
              return cfg.color;
            })
          }]
        },
        options: {
          plugins: {
            legend: {display: false},
            title: {
              display: true,
              text: [cfg.titulo, PERIODO.etiqueta || ""],
              font: {size: 13}, padding: {bottom: 12}
            },
            tooltip: {callbacks: {
              title: function (ctx) {
                return etiquetaLarga(filas[ctx[0].dataIndex]);
              },
              label: function (ctx) {
                return cfg.eje + ": " + cfg.formato(ctx.raw);
              },
              afterLabel: function (ctx) {
                const v = filas[ctx.dataIndex];
                return [
                  "Entregas: " + SL.entero(v.entregas),
                  "Costo total: " + SL.dinero(v.costo_total),
                  "Retraso medio: " + (v.retraso_medio_min === null
                    ? "—" : SL.numero(v.retraso_medio_min, 1) + " min")
                ];
              }
            }}
          },
          scales: {
            x: {title: {display: true, text: "Vehículo"}},
            y: {
              beginAtZero: criterio !== "rendimiento",
              title: {display: true, text: cfg.eje},
              ticks: {callback: function (v) {
                return criterio === "costo" ? SL.dinero(v) : SL.numero(v, 0);
              }}
            }
          }
        }
      });
    } catch (error) {
      document.getElementById("l-flotilla").textContent = error.message;
    }
  }

  // ======================================================================
  // COSTO FRENTE A TRABAJO
  // ======================================================================
  async function comparativa() {
    try {
      const r = await SL.api("/analitica/vehiculos?orden=costo&top=100");
      const d = r.datos;
      flotillaCompleta = d.vehiculos;
      pintarTabla(flotillaCompleta);

      const maxLitros = Math.max.apply(null,
        flotillaCompleta.map(function (v) { return v.litros; })) || 1;
      const costoMedio = d.totales.costo_medio_por_vehiculo;
      const entregasMedia = d.totales.entregas / (d.flotilla || 1);

      // Ineficientes = por encima del costo medio y por debajo de las
      // entregas medias. Es el cuadrante que interesa, no el ranking.
      const ineficientes = flotillaCompleta.filter(function (v) {
        return v.costo_total > costoMedio && v.entregas < entregasMedia;
      });

      if (dispersion) dispersion.destroy();
      dispersion = new Chart(document.getElementById("g-dispersión"), {
        type: "bubble",
        data: {
          datasets: [
            {
              label: "Cuesta más de lo que trabaja",
              data: ineficientes.map(function (v) {
                return {x: v.entregas, y: v.costo_total,
                        r: 4 + 12 * v.litros / maxLitros, v: v};
              }),
              backgroundColor: "rgba(214,39,40,.65)"
            },
            {
              label: "Dentro de lo esperado",
              data: flotillaCompleta.filter(function (v) {
                return ineficientes.indexOf(v) < 0;
              }).map(function (v) {
                return {x: v.entregas, y: v.costo_total,
                        r: 4 + 12 * v.litros / maxLitros, v: v};
              }),
              backgroundColor: "rgba(31,78,121,.55)"
            }
          ]
        },
        options: {
          plugins: {
            title: {
              display: true,
              text: ["Costo de operación frente a entregas realizadas",
                     PERIODO.etiqueta || ""],
              font: {size: 13}, padding: {bottom: 12}
            },
            legend: {position: "bottom", labels: {boxWidth: 10}},
            tooltip: {callbacks: {
              label: function (ctx) {
                const v = ctx.raw.v;
                return [
                  etiquetaLarga(v),
                  "Entregas: " + SL.entero(v.entregas),
                  "Costo: " + SL.dinero(v.costo_total),
                  "Costo por entrega: " + SL.dinero(v.costo_por_entrega),
                  "Combustible: " + SL.numero(v.litros, 0) + " litros"
                ];
              }
            }}
          },
          scales: {
            x: {title: {display: true, text: "Entregas realizadas"},
                beginAtZero: true},
            y: {title: {display: true, text: "Costo de operación (MXN)"},
                beginAtZero: true,
                ticks: {callback: function (v) { return SL.dinero(v); }}}
          }
        }
      });

      const zona = document.getElementById("l-dispersion");
      if (ineficientes.length) {
        const peor = ineficientes.slice().sort(function (a, b) {
          return b.costo_por_entrega - a.costo_por_entrega;
        })[0];
        zona.textContent =
          ineficientes.length + " de las " + d.flotilla + " unidades gastan " +
          "por encima de la media (" + SL.dinero(costoMedio) + ") y entregan " +
          "por debajo de ella (" + SL.numero(entregasMedia, 0) + " entregas). " +
          "La más desproporcionada es " + etiquetaLarga(peor) + ", con " +
          SL.dinero(peor.costo_por_entrega) + " por entrega frente a los " +
          SL.dinero(d.totales.costo_total / d.totales.entregas) +
          " del promedio de la flotilla.";
      } else {
        zona.textContent =
          "Ninguna unidad gasta por encima de la media entregando por debajo " +
          "de ella: el costo de la flotilla acompaña al trabajo que hace cada " +
          "vehículo.";
      }
    } catch (error) {
      document.getElementById("l-dispersion").textContent = error.message;
    }
  }

  // ======================================================================
  // TABLA
  // ======================================================================
  function pintarTabla(filas) {
    const cuerpo = document.getElementById("tabla-flotilla");
    cuerpo.innerHTML = "";
    filas.forEach(function (v) {
      const tarde = v.retraso_medio_min !== null && v.retraso_medio_min > UMBRAL;
      const flojo = (v.desviacion_rendimiento_pct || 0) < 0;
      const tr = document.createElement("tr");
      tr.innerHTML =
        "<td><strong>" + SL.escapar(v.codigo_vehiculo) + "</strong>" +
          "<br><small class='text-body-secondary'>" +
          SL.escapar(v.descripcion) + " · " + SL.escapar(v.placa) +
        "</small></td>" +
        "<td>" + SL.estado(v.estado_operativo) + "</td>" +
        '<td class="sl-num">' + SL.entero(v.viajes) + "</td>" +
        '<td class="sl-num">' + SL.entero(v.entregas) + "</td>" +
        '<td class="sl-num">' + SL.numero(v.km_recorridos, 0) + "</td>" +
        '<td class="sl-num">' + SL.dinero(v.costo_total) + "</td>" +
        '<td class="sl-num">' + SL.dinero(v.costo_por_entrega) + "</td>" +
        '<td class="sl-num">' + SL.numero(v.litros, 0) + "</td>" +
        '<td class="sl-num ' + (flojo ? "text-danger fw-semibold" : "") + '">' +
          SL.numero(v.rendimiento_real_km_l, 2) +
          "<br><small class='text-body-secondary'>ficha " +
          SL.numero(v.rendimiento_nominal_km_l, 2) + "</small></td>" +
        '<td class="sl-num ' + (tarde ? "text-danger fw-semibold" : "") + '">' +
          (v.retraso_medio_min === null ? "—" :
            SL.numero(v.retraso_medio_min, 1)) +
          "<br><small class='text-body-secondary'>" +
          (v.pct_retrasadas === null ? "" :
            SL.numero(v.pct_retrasadas, 0) + "% tarde") + "</small></td>" +
        '<td class="sl-num">' + SL.entero(v.mantenimientos) + "</td>";
      cuerpo.appendChild(tr);
    });
  }

  // ======================================================================
  // MANTENIMIENTO PENDIENTE
  // ======================================================================
  async function mantenimiento() {
    const zona = document.getElementById("mantenimiento");
    try {
      const r = await SL.api("/mantenimientos/pendientes");
      const d = r.datos;
      const filas = d.vencidos.concat(d.atrasados, d.proximos);

      if (!filas.length) {
        zona.className = "sl-vacio";
        zona.innerHTML = '<i class="bi bi-check-circle"></i>' +
          "<h3>Ninguna unidad pendiente</h3><p></p>";
        zona.querySelector("p").textContent = d.alerta;
        return;
      }

      const etiquetas = {vencidos: ["Vencido", "sl-pastilla-mal"],
                         atrasados: ["Atrasado", "sl-pastilla-alerta"],
                         proximos: ["Próximo", "sl-pastilla-info"]};
      let html = '<p class="sl-lectura mt-0 mb-3">' + SL.escapar(d.alerta) +
                 "</p>";
      html += '<div class="sl-tabla-envoltura"><table class="table table-sm ' +
              'sl-tabla mb-0"><thead><tr><th>Vehículo</th><th>Situación</th>' +
              "<th>Tipo</th><th>Fecha programada</th>" +
              '<th class="text-end">Días de diferencia</th>' +
              "<th>Estado de la unidad</th></tr></thead><tbody>";

      ["vencidos", "atrasados", "proximos"].forEach(function (grupo) {
        d[grupo].forEach(function (m) {
          const et = etiquetas[grupo];
          html += "<tr><td><strong>" + SL.escapar(m.codigo_vehiculo) +
            "</strong> <small class='text-body-secondary'>" +
            SL.escapar(m.placa || "") + "</small></td>" +
            '<td><span class="sl-pastilla ' + et[1] + '">' + et[0] +
            "</span></td>" +
            "<td>" + SL.estado(m.tipo) + "</td>" +
            "<td>" + SL.fecha(m.fecha_programada) + "</td>" +
            '<td class="sl-num">' + SL.entero(m.dias) + "</td>" +
            "<td>" + SL.estado(m.estado_operativo) + "</td></tr>";
        });
      });
      html += "</tbody></table></div>";

      zona.className = "";
      zona.innerHTML = html;
    } catch (error) {
      zona.textContent = error.message;
    }
  }

  document.getElementById("criterio").addEventListener("change", cargar);
  document.getElementById("cuantos").addEventListener("change", cargar);

  if (document.getElementById("g-flotilla")) {
    cargar();
    comparativa();
    mantenimiento();
  }
})();
